master/internal/experiment/bulk_action.go

```diff
@@ -734,3 +735,90 @@ func MoveExperiments(ctx context.Context,
 	}
 	return results, nil
 }
+
+func changeExperimentConfigLogRetention(ctx context.Context, database db.DB,
+	expID int, numDays int16,
+) error {
+	exp, err := db.ExperimentByID(ctx, expID)
+	if err != nil {
+		return errors.Wrap(err, "fetching experiment from database")
+	}
+
+	if !model.TerminalStates[exp.State] {
+		return status.Error(codes.FailedPrecondition, fmt.Sprintf(
+			"experiment in non terminal state '%s', try again later", exp.State))
+	}
+
+	activeConfig, err := database.ActiveExperimentConfig(exp.ID)
+	if err != nil {
+		return errors.Wrapf(
+			err, "unable to load config for experiment %v", exp.ID,
+		)
+	}
+	activeConfig.SetLogRetentionDays(&numDays)
+
+	if err := database.SaveExperimentConfig(exp.ID, activeConfig); err != nil {
+		return errors.Wrapf(err, "patching experiment %d", exp.ID)
+	}
+
+	return nil
+}
+
+// RetainLogs retains logs for the given list of experiments.
+func RetainLogs(ctx context.Context, database db.DB,
```

Comment:
it's not clear to me why this needs to only operate on active experiments (but i remember we talked about this last week, lol). maybe this needs some commentary.

also, I remember we talked and I can't remember the conclusion, but don't remember it being this? i remember you had a good idea that was much less work. curious, what changed?

Reply:
Actually, the functionality I am going for is to not allow for this to operate on active experiments, only experiments in terminal state.

The issue was that in long experiments like asha, if the user called this api right after creating the experiment, only the trials that were created at that time have the log retention days changed, any trials created after a while will have the same log retention days as in the task spec.

We talked about trying to change the task spec while the experiment is running, or having some sort of wait that will change the log retention days of the trials after the experiment is completed but we came to the conclusion that it would be simplest to not allow the users to run this command before an experiment is completed.

Reply:
Ah yeah, i'm recalling now. You should put this all in a comment somewhere, it is useful!

Reply:
done!